### PR TITLE
perf: during file update request only changed portion of the text from the …

### PR DESCRIPTION
**Pull request by vladima**
…host

### CODE REVIEW

File: src/services/services.ts [hunk 1]
```diff
@@ -1829,7 +1829,34 @@ module ts {
...
-                    let newSourceFile = updateSourceFile(sourceFile, scriptSnapshot.getText(0, scriptSnapshot.getLength()), textChangeRange, aggressiveChecks);
+                    let newText: string;
+                    
+                    // grab the fragment from the beginning of the original text to the beginning of the span
+                    let prefix = textChangeRange.span.start !== 0
+                        ? sourceFile.text.substr(0, textChangeRange.span.start)
+                        : "";
+                    
+                    // grab the fragment from the end of the span till the end of the original text
+                    let suffix = textChangeRange.span.start + textChangeRange.span.length !== sourceFile.text.length
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Can you factor `textChangeRange.span.start + textChangeRange.span.length` to `textChangeRangeEnd`?

File: src/services/services.ts [hunk 2]
```diff
@@ -1829,7 +1829,34 @@ module ts {
...
+
+                    if (textChangeRange.newLength === 0) {
+                        // edit was a deletion - just combine prefix and suffix
+                        newText = prefix && suffix ? prefix + suffix : prefix || suffix;
+                    }
+                    else {
+                        // it was actual edit, fetch the fragment of new text that correspond to new span
+                        let changedText = scriptSnapshot.getText(textChangeRange.span.start, textChangeRange.span.start + textChangeRange.newLength);
+                        // combine prefix, changed text and suffix
+                        newText = prefix && suffix
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Are we not feeling confident enough that `""  + blah + ""` ends up returning `blah`?

Comment (vladima) [inline src/services/services.ts]: I'd prefer to keep the amount of string concatenations as small as possible

File: src/services/services.ts [hunk 3]
```diff
@@ -1829,7 +1829,35 @@ module ts {
...
                 if (!disableIncrementalParsing) {
-                    let newSourceFile = updateSourceFile(sourceFile, scriptSnapshot.getText(0, scriptSnapshot.getLength()), textChangeRange, aggressiveChecks);
+                    let newText: string;
+                    
+                    // grab the fragment from the beginning of the original text to the beginning of the span
+                    let prefix = textChangeRange.span.start !== 0
+                        ? sourceFile.text.substr(0, textChangeRange.span.start)
+                        : "";
+                    
+                    let textChangeRangeEnd = textChangeRange.span.start + textChangeRange.span.length;
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: you can say spanEnd(textChangeRange.span);

File: src/services/services.ts [hunk 4]
```diff
@@ -1829,7 +1829,35 @@ module ts {
...
+                    // grab the fragment from the end of the span till the end of the original text
+                    let suffix = textChangeRangeEnd !== sourceFile.text.length
+                            ? sourceFile.text.substr(textChangeRangeEnd)
+                            : "";
+
+                    if (textChangeRange.newLength === 0) {
+                        // edit was a deletion - just combine prefix and suffix
+                        newText = prefix && suffix ? prefix + suffix : prefix || suffix;
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: Cute.  But ideally unnecessary.  I would hope that the engines would optimize concats of zero-length strings to be no-ops with no allocations.

File: src/services/services.ts (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/services/services.ts]: i'm skeptical we need to go to this degree of over optimization.  This really only helps when you're literally at the start or the end of the file.  That's such a narrow case.  And presumes hte engines aren't already smart enough to do the right thing here.

**Comment by CyrusNajmabadi** [comment]
:+1:

File: src/services/services.ts (hunk 2 above)
Comment (vladima) [inline src/services/services.ts]: based on the presentation that  we've seen recently I can reliably say only that I cannot trust engines at all.

File: src/services/services.ts (hunk 3 above)
Comment (vladima) [inline src/services/services.ts]: will do

File: src/services/services.ts [hunk 5]
```diff
@@ -1829,7 +1829,34 @@ module ts {
...
-                    let newSourceFile = updateSourceFile(sourceFile, scriptSnapshot.getText(0, scriptSnapshot.getLength()), textChangeRange, aggressiveChecks);
+                    let newText: string;
+                    
+                    // grab the fragment from the beginning of the original text to the beginning of the span
+                    let prefix = textChangeRange.span.start !== 0
+                        ? sourceFile.text.substr(0, textChangeRange.span.start)
+                        : "";
+                    
+                    // grab the fragment from the end of the span till the end of the original text
+                    let suffix = textSpanEnd(textChangeRange.span) !== sourceFile.text.length
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Still odd that it's called twice but I'm fine with this.

**Comment by DanielRosenwasser** [comment]
:+1:

**Pull request merged**